File: Ad-Hoc/__3000__/3206/3206.py
class Robot:
    def __init__(self, x, y, sense):
        self.x = x
        self.y = y
        self.sense = sense

    def move(self):
        deltas = {
            'W': (-1, 0),
            'N': (0, 1),
            'E': (1, 0),
            'S': (0, -1)
        }

        self.x += deltas[self.sense][0]
        self.y += deltas[self.sense][1]

    def rotate(self, shift):
        senses = "NESW"
        idx = senses.index(self.sense)
        self.sense = senses[(idx + shift + 4) % 4]

    def execute(self, command):
        if command == 'F':
            self.move()
        elif command == 'L':
            self.rotate(-1)
        elif command == 'R':
            self.rotate(1)


T = int(input())

for _ in range(T):
    width, height = map(int, input().split())
    nrobots, ncmds = map(int, input().split())

    robots = []
    grid = [[0] * 105 for _ in range(105)]

    for robot_id in range(nrobots):
        x, y, sense = input().split()
        x = int(x)
        y = int(y)

        robots.append(Robot(x, y, sense))
        grid[x][y] = robot_id + 1  # IDs são 1-based

    collided = False
    ans = "OK"

    for _ in range(ncmds):
        ID, cmd, rep = input().split()
        ID = int(ID)
        rep = int(rep)

        if collided:
            continue

        robot = robots[ID - 1]

        # rotações não precisam consultar grid
        if cmd == 'L':
            robot.rotate(-(rep % 4))
            continue

        if cmd == 'R':
            robot.rotate(rep % 4)
            continue

        # cmd == 'F'
        for _ in range(rep):
            cx, cy = robot.x, robot.y

            robot.move()

            x, y = robot.x, robot.y

            # colisão com parede
            if x <= 0 or x > width or y <= 0 or y > height:
                collided = True
                ans = f"Robot {ID} crashes into the wall"
                break

            # colisão com outro robô
            if grid[x][y] != 0 and grid[x][y] != ID:
                collided = True
                ans = f"Robot {ID} crashes into robot {grid[x][y]}"
                break

            # atualiza grid
            grid[cx][cy] = 0
            grid[x][y] = ID

    print(ans)

File: Ad-Hoc/__3000__/3206/3206.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin", "utf8").trim().split("\n");
let idx = 0;

let T = Number(input[idx++]);
let output = [];

while (T--) {
  const grid = Array.from({ length: 105 }, () => Array(105).fill(0));

  const width = Number(input[idx++]);
  const height = Number(input[idx++]);

  const numRobots = Number(input[idx++]);
  let numCmd = Number(input[idx++]);

  const robots = [];

  for (let i = 0; i < numRobots; i++) {
    const x = Number(input[idx++]);
    const y = Number(input[idx++]);
    const sense = input[idx++];

    robots.push(new Robot(x, y, sense));
    grid[x][y] = i + 1; // ID 1-based
  }

  let collided = false;
  let ans = "OK\n";

  while (numCmd--) {
    const id = Number(input[idx++]);
    const cmd = input[idx++];
    const reps = Number(input[idx++]);

    if (collided) continue;

    const robot = robots[id - 1];

    for (let r = 0; r < reps; r++) {
      const cx = robot.x;
      const cy = robot.y;

      // turn left
      if (cmd === 'L') {
        changeSense(robot, -1);
        continue;
      }

      // turn right
      if (cmd === 'R') {
        changeSense(robot, 1);
        continue;
      }

      // move forward
      if (cmd === 'F') {
        moveRobot(robot);

        // wall collision
        if (robot.x <= 0 || robot.x > width || robot.y <= 0 || robot.y > height) {
          collided = true;
          ans = `Robot ${id} crashes into the wall\n`;
          break;
        }

        // robot collision
        if (grid[robot.x][robot.y] !== 0 && grid[robot.x][robot.y] !== id) {
          collided = true;
          ans = `Robot ${id} crashes into robot ${grid[robot.x][robot.y]}\n`;
          break;
        }

        // update grid
        grid[cx][cy] = 0;
        grid[robot.x][robot.y] = id;
      }
    }
  }

  output.push(ans);
}

console.log(output.join(""));